File: src/main/java/app/application/utils/ExceptionExtractor.java
package app.application.utils;

import org.springframework.stereotype.Service;

import java.lang.reflect.InvocationTargetException;

@Service
public class ExceptionExtractor {

	public Throwable getTargetOfException(Exception exception){
		exception = (Exception) exception.getCause();
		if(!isTargetException(exception)){
			return exception.getCause();
		}
		return ((InvocationTargetException) exception).getTargetException();
	}

	private boolean isTargetException(Exception exception){
		return exception instanceof InvocationTargetException;
	}

}

File: src/main/kotlin/app/application/utils/ExceptionExtractor.kt
package app.application.utils

import org.springframework.stereotype.Service
import java.lang.reflect.InvocationTargetException

@Service
class ExceptionExtractor {

    fun getTargetOfException(exception: Exception?): Throwable? {
        val innerException: Exception? = exception?.cause as Exception?
        return if (!isTargetException(innerException)) {
            innerException?.cause
        } else (innerException as InvocationTargetException?)?.targetException
    }

    private fun isTargetException(exception: Exception?): Boolean {
        return exception is InvocationTargetException
    }
}